### ISSUE TITLE
104.ua API

### ISSUE THREAD

**Issue by 5kull** [Krezalis/metersen#1]
Привіт,
Якщо можеш поясни будь ласка стосовно відправки показників газу по АРІ на 104.ua

"https://mobile.104.ua/billing/api/v2/users/username@gmail.com/sessions?password=12345678&device_id=1 У відповідь отримуємо data:session_id Передавати заголовки: X-Application-Key: витягуємо із apk файлу X-Session-Id: session_id"

- Як витягнути X-Application-Key, треба скачати з телефону вже встановлену АРК і декомпілити її чимось?
і взагалі цю частину покроково якщо можна, я так розумію ми маємо отримати тут якийсь токен для подальших дій?

**Comment by Krezalis** [comment]
так.
З цікавого у 2.0.17 новий ключ додатку.
По апі каша, 2.*.* працює із версіями апі в перемішку v2, v3, v4. У v4 реалізовано передачу фото у POST запиті для розпізнавання.

**Comment by 5kull** [comment]
Підкажи будь ласка що не так, ось мій запрос (логін пароль змінені)
`curl -i -H "Accept: application/json" -H "X-Application-Key: com.rgc.client" https://mobile.104.ua/billing/api/v2/users/mail@gmail.com/sessions?password=123456&device_id=1`

Отримую:
{"status_code":100000,"status_message":"Unknown error"}

**Comment by Krezalis** [comment]
- тут має бути POST тому бє 500 помилку
- після цього стикнетеся із 401 помилкою

**Comment by sa-egor** [comment]
Сейчас начал тут смотреть их новый кабинет... У меня, вроде как (потому что сначала с сайта отправил, а теперь через запросы ошибка 375796 о том, что показания ещё в обработке), получилось отправить так:
### Авторизация:
**URL**: _https://ok.104.ua/ua/ajx/individual/signin/auth/credentials_
**Запрос**: POST
**Данные формы**: _login=mail@example.com&password=mysecurepass_

В ответ приходят кукисы PHPSESSID и device_view. Я пока не совсем понял нужно ли сохранять device_view или нет, но сейчас через PHP сохраняю всё.

### Отправка показаний:
**URL**: _https://ok.104.ua/ua/ajx/individual/meterage/send_
**Запрос**: POST
**Данные формы**: _account_no=0000000000&meter_no=0000000&value=123_
Где:
account_no - личный счет. Отображается на сайте слева сверху.
meter_no - id счетчика. Найти можно по ссылке: https://ok.104.ua/ua/meterage
value - показания счётчика. То есть цифры с него.

В PHP у меня это выглядит так:

```
$ckfile = tempnam ("/tmp", 'cookiename');
$login_url = "https://ok.104.ua/ua/ajx/individual/signin/auth/credentials";
$send_data_url = "https://ok.104.ua/ua/ajx/individual/meterage/send";

$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, $login_url);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
curl_setopt($ch, CURLOPT_POST, true);
curl_setopt ($ch, CURLOPT_COOKIEJAR, $ckfile);
curl_setopt ($ch, CURLOPT_COOKIEFILE, $ckfile);
curl_setopt($ch, CURLOPT_POSTFIELDS, "login=mail@example.com&password=mysecurepass");
$output = curl_exec($ch);

$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, $send_data_url);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
curl_setopt($ch, CURLOPT_POST, true);
curl_setopt ($ch, CURLOPT_COOKIEJAR, $ckfile);
curl_setopt ($ch, CURLOPT_COOKIEFILE, $ckfile);
curl_setopt($ch, CURLOPT_POSTFIELDS, "account_no=0000000000&meter_no=0000000&value=123");
$output = curl_exec($ch);

var_dump($output);
print("<br><br>");
var_dump($ckfile);
```

2 var_dump в конце чисто чтобы увидеть ответ. В случае если в этом нет необходимости, то можно их удалить.
Я знаю что необзательно инициализировать курл ещё раз, это чисто быстрая попытка. Потом причешу у себя этот код.

**Comment by Krezalis** [comment]
> Сейчас начал тут смотреть их новый кабинет... У меня, вроде как (потому что сначала с сайта отправил, а теперь через запросы ошибка 375796 о том, что показания ещё в обработке), получилось отправить так:

Вебівський є небезпека що додадуть рекапчу. Із апі мобільного додатку важче додати захист від ботів)

**Comment by sa-egor** [comment]
Согласен. Но и на мобильном могут капчу добавить. В любом случае пока попользуюсь этим, а дальше видно будет)

**Comment by Krezalis** [comment]
Куди варто подивитися: проект ватеріус переробити під 104mobile, буде в пару разів дешевше ...

**Comment by sa-egor** [comment]
Ну такое устройство собрать без проблем. На час работы)
У нас тут получается больше проблем именно с отправкой. То есть пока 104 ещё не закончили модификации сайта и приложения, то всё может измениться не один раз. Потому тут или ждать от них АПИ, или, условно говоря, каждый месяц менять принцип отправки в своих скриптах, когда они что-то меняют у себя.
